sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala

```diff
@@ -39,7 +39,19 @@ private[sql] object JsonRDD extends Logging {
       json: RDD[String],
       schema: StructType,
       columnNameOfCorruptRecords: String): RDD[Row] = {
-    parseJson(json, columnNameOfCorruptRecords).map(parsed => asRow(parsed, schema))
+    // Reuse the mutable row for each record, however we still need to 
+    // create a new row for every nested struct type in each record
+    val mutableRow = new SpecificMutableRow(schema.fields.map(_.dataType))
+    parseJson(json, columnNameOfCorruptRecords).mapPartitions( iter => {
+      iter.map { parsed =>
+        schema.fields.zipWithIndex.foreach {
```

Comment:
This is duplicated with the function `asRow`, can we add additional parameter for `asRow`, says
```
def asRow(json: Map[String,Any], schema: StructType, mutable: GenericMutableRow = null): Row = {
  row = if (mutable == null) {
     new GenericMutableRow(schema.fields.length)
  } else {
    mutable
  }
}
```